src/Rules/Register.hs

```diff
@@ -1,44 +1,101 @@
-        "//" ++ stage0PackageDbDir -/- pkgName package ++ "*.conf" %%>
-            buildConf rs context
+registerPackages :: [(Resource, Int)] -> Context -> Rules ()
+registerPackages rs context@Context {..} = do
+    root <- buildRootRules
+    root -/- inplacePackageDbPath stage %>
+      buildStamp rs context
+
+    root -/- inplacePackageDbPath stage -/- packageDbStamp %> \stamp -> do
+      writeFileLines stamp []
 
-        when (package == ghc) $ "//" ++ stage0PackageDbDir -/- packageDbStamp %>
-            buildStamp rs context
+    root -/- inplacePackageDbPath stage -/- "*.conf" %> \conf -> do
+      settings <- libPath context <&> (-/- "settings")
+      platformConstants <- libPath context <&> (-/- "platformConstants")
 
-    when (stage == Stage1) $ do
-        inplacePackageDbPath -/- pkgName package ++ "*.conf" %%>
-            buildConf rs context
+      need [settings, platformConstants]
+      let Just pkgName | takeBaseName conf == "rts" = Just "rts"
+                       | otherwise = fst <$> parseCabalName (takeBaseName conf)
+      let Just pkg = findPackageByName pkgName
+      bootLibs <- filter isLibrary <$> (defaultPackages Stage0)
+      case stage of
+        Stage0 | not (pkg `elem` bootLibs) -> copyConf rs (context { package = pkg }) conf
+        _                                  -> buildConf rs (context { package = pkg }) conf
 
-        when (package == ghc) $ inplacePackageDbPath -/- packageDbStamp %>
-            buildStamp rs context
+copyConf :: [(Resource, Int)] -> Context -> FilePath -> Action ()
+copyConf rs context@Context {..} conf = do
+    depPkgIds <- fmap stdOutToPkgIds . askWithResources rs $
+      target context (GhcPkg Dependencies stage) [pkgName package] []
+    need =<< mapM (\pkgId -> packageDbPath stage <&> (-/- pkgId <.> "conf")) depPkgIds
+    buildWithResources rs $ do
+      target context (GhcPkg Clone stage) [pkgName package] [conf]
+
+  where
+    stdOutToPkgIds :: String -> [String]
+    stdOutToPkgIds = drop 1 . concatMap words . lines
 
 buildConf :: [(Resource, Int)] -> Context -> FilePath -> Action ()
-buildConf rs context@Context {..} conf = do
-    confIn <- pkgInplaceConfig context
-    need [confIn]
-    buildWithResources rs $ target context (GhcPkg Update stage) [confIn] [conf]
+buildConf _ context@Context {..} _conf = do
+
+    depPkgIds <- cabalDependencies context
+
+    -- setup-config, triggers `ghc-cabal configure`
+    -- everything of a package should depend on that
+    -- in the first place.
+    setupConfig <- (contextPath context) <&> (-/- "setup-config")
+    need [setupConfig]
+
+    need =<< mapM (\pkgId -> packageDbPath stage <&> (-/- pkgId <.> "conf")) depPkgIds
+
+    ways <- interpretInContext context (getLibraryWays <> if package == rts then getRtsWays else mempty)
+    need =<< concatMapM (libraryTargets True) [ context { way = w } | w <- ways ]
+
+    -- might need some package-db resource to limit read/write,
```

Comment:
"might" meaning that this needs to be implemented? Hadrian used to do this, no?

Reply:
This used to be in a different action. Probably stale. However this action, while it gets the resource info, discards that argument. Not sure if we need to use it here.